src/Workspaces/Core/Portable/Shared/Extensions/SourceTextExtensions.cs

```diff
@@ -248,30 +259,25 @@ public static TextReader Create(ObjectReader reader)
                 return new StringReader(reader.ReadString());
             }
 
-            // read as chunks
-            using var _ = ArrayBuilder<char[]>.GetInstance(out var builder);
-
             var chunkSize = reader.ReadInt32();
             var numberOfChunks = reader.ReadInt32();
 
+            // read as chunks
+            using var _ = ArrayBuilder<char[]>.GetInstance(numberOfChunks, out var chunks);
+
             var offset = 0;
             for (var i = 0; i < numberOfChunks; i++)
             {
-                var (currentChunk, currentChunkLength) = reader.ReadCharArray(static length =>
-                {
-                    if (length <= SharedPools.CharBufferSize)
-                        return SharedPools.CharArray.Allocate();
-
-                    return new char[length];
-                });
-
-                builder.Add(currentChunk);
+                // Shared pool array will be freed in the Dispose method below.
+                var (currentChunk, currentChunkLength) = reader.ReadCharArray(
+                    static length => length == CharArrayLength ? s_charArrayPool.Allocate() : new char[length]);
```

Comment:
>static length => length == CharArrayLength ? s_charArrayPool.Allocate() : new char[length]) [](http://example.com/codeflow?start=20&length=91)

For small buffers (< 4KB), is the allocation profile worse now?

Reply:
Ah. Good point. I'll fix.